### PR TITLE
Adding support for 'env' config option

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

Adding support for 'env' config option.
It represents the environment name were the agent is running and will be attached in-app to every metric, event, log, trace, and service check emitted by this Agent.

### CODE REVIEW

File: pkg/config/config_template.yaml [hunk 1]
```diff
@@ -65,6 +65,12 @@ api_key:
 #   - environment:dev
 #   - <TAG_KEY>:<TAG_VALUE>
 
+## @param end  - string - optional
+## The environment name were the agent is running. Attached in-app to every
+## metric, event, log, trace, and service check emitted by this Agent.
+#
+# env: <environment name>
```
Comment (l0k0ms) [inline pkg/config/config_template.yaml]:
```suggestion
# env: <ENVIRONMENT_NAME>
```

Comment (l0k0ms) [reply pkg/config/config_template.yaml]: Shouldn't we set a default `dev` or `test` value ?

File: pkg/config/config_template.yaml [hunk 2]
```diff
@@ -65,6 +65,12 @@ api_key:
 #   - environment:dev
 #   - <TAG_KEY>:<TAG_VALUE>
 
+## @param end  - string - optional
+## The environment name were the agent is running. Attached in-app to every
```
Comment (olivielpeau) [inline pkg/config/config_template.yaml]:
```suggestion
## The environment name where the agent is running. Attached in-app to every
```

File: pkg/config/config_template.yaml (hunk 1 above)
Comment (olivielpeau) [reply pkg/config/config_template.yaml]: (the default should be empty, i.e. if the user doesn't specify anything, the agent should not send the `env:` host tag)

**Review by olivielpeau — COMMENTED** [review]
let's update the trace agent to pick up this new option as well

File: pkg/config/config_template.yaml [hunk 3]
```diff
@@ -65,6 +65,12 @@ api_key:
 #   - environment:dev
 #   - <TAG_KEY>:<TAG_VALUE>
 
+## @param end  - string - optional
```
Comment (gbbr) [inline pkg/config/config_template.yaml]:
```suggestion
## @param env - string - optional
```

File: pkg/config/config_template.yaml [hunk 4]
```diff
@@ -445,7 +451,8 @@ api_key:
 
   ## @param env - string - optional - default: none
   ## The environment tag that Traces should be tagged with.
-  ## Inherits from "env" tag if "none" is applied here.
+  ## If "none" is set here the value will be inherited first from the top level
```
Comment (gbbr) [inline pkg/config/config_template.yaml]:
```sh
  ## If left empty or set to "none", the value will be inherited from the top level
  ## "env" config option, if set.
```
Maybe this is enough? And then, we can mention on the top level "env" option that the `env:` tag will override this setting.

File: pkg/trace/config/apply.go [hunk 5]
```diff
@@ -173,9 +173,23 @@ func (c *AgentConfig) applyDatadogConfig() error {
 	if config.Datadog.IsSet("apm_config.log_file") {
 		c.LogFilePath = config.Datadog.GetString("apm_config.log_file")
 	}
+
 	if config.Datadog.IsSet("apm_config.env") {
 		c.DefaultEnv = config.Datadog.GetString("apm_config.env")
+		log.Debugf("setting DefaultEnv to '%s' (from apm_config.env)", c.DefaultEnv)
```
Comment (gbbr) [inline pkg/trace/config/apply.go]: I think @mfpierre made it a "convention" to always capitalise log messages. Can we do that please?
```suggestion
		log.Debugf("Setting DefaultEnv to %q (from apm_config.env)", c.DefaultEnv)
```

File: pkg/trace/config/apply.go [hunk 6]
```diff
@@ -173,9 +173,23 @@ func (c *AgentConfig) applyDatadogConfig() error {
...
+	} else if config.Datadog.IsSet("env") {
+		c.DefaultEnv = config.Datadog.GetString("env")
+		log.Debugf("setting DefaultEnv to '%s' (from 'env' config option)", c.DefaultEnv)
+	} else if config.Datadog.IsSet("tags") {
+		for _, tag := range config.Datadog.GetStringSlice("tags") {
+			if strings.HasPrefix(tag, "env:") {
+				c.DefaultEnv = strings.SplitN(tag, "env:", 2)[1]
```
Comment (gbbr) [inline pkg/trace/config/apply.go]: This works, but is slightly scary (thinking of panic, even if not possible in the current form). Can we instead do:
```suggestion
				c.DefaultEnv = strings.TrimPrefix(tag, "env:")
```

File: pkg/trace/config/apply.go [hunk 7]
```diff
@@ -173,9 +173,23 @@ func (c *AgentConfig) applyDatadogConfig() error {
...
+				log.Debugf("setting DefaultEnv to '%s' (from host tag: %s)", c.DefaultEnv, tag)
```
Comment (gbbr) [inline pkg/trace/config/apply.go]:
```suggestion
				log.Debugf("Setting DefaultEnv to %q (from host tag: %q)", c.DefaultEnv, tag)
```
This will add quotes automagically.

File: pkg/trace/config/apply.go [hunk 8]
```diff
@@ -173,9 +173,23 @@ func (c *AgentConfig) applyDatadogConfig() error {
 	if config.Datadog.IsSet("apm_config.log_file") {
 		c.LogFilePath = config.Datadog.GetString("apm_config.log_file")
 	}
+
 	if config.Datadog.IsSet("apm_config.env") {
 		c.DefaultEnv = config.Datadog.GetString("apm_config.env")
+		log.Debugf("setting DefaultEnv to '%s' (from apm_config.env)", c.DefaultEnv)
+	} else if config.Datadog.IsSet("env") {
+		c.DefaultEnv = config.Datadog.GetString("env")
+		log.Debugf("setting DefaultEnv to '%s' (from 'env' config option)", c.DefaultEnv)
```
Comment (gbbr) [inline pkg/trace/config/apply.go]:
```suggestion
		log.Debugf("Setting DefaultEnv to %q (from 'env' config option)", c.DefaultEnv)
```

File: pkg/trace/config/apply.go [hunk 9]
```diff
@@ -173,9 +173,23 @@ func (c *AgentConfig) applyDatadogConfig() error {
 	if config.Datadog.IsSet("apm_config.log_file") {
 		c.LogFilePath = config.Datadog.GetString("apm_config.log_file")
 	}
+
```
Comment (gbbr) [inline pkg/trace/config/apply.go]: APM styling nit, but can we remove the empty lines before and after the `if` block? I find that generally the curly braces provide enough spacing, and I've also noticed that in Go spacing like this to create "paragraphs of code" is not really used much looking at larger projects and the `golang/go` repository itself.

**Review by gbbr — COMMENTED** [review]
Thanks for doing this work @hush-hush. Left a few nits.

File: pkg/config/config_template.yaml (hunk 4 above)
Comment (hush-hush) [reply pkg/config/config_template.yaml]: the `env:` tag will only overwrite it for trace, the agent will send both. I think we have to explain it.

Comment (gbbr) [reply pkg/config/config_template.yaml]: Oh, right. That’s true. Maybe we can phrase it a bit better? The current one is a bit hard to understand. Maybe @l0k0ms can suggest something as the docs expert.

**Review by gbbr — DISMISSED** [review]
This LGTM as far as trace-agent changes are concerned. Thanks Maxime 🙏🏻

**Review by gbbr — DISMISSED** [review]

File: pkg/config/config_template.yaml (hunk 4 above)
Comment (olivielpeau) [reply pkg/config/config_template.yaml]: It's a bit of a weird edge case, but if I set the following in the config:

```yaml
env: "foo"
apm_config:
  env: "none"
```

based on the current documentation here, I'd expect the APM Agent to pick up `foo` as the env tag. But it I understand the code correctly, the APM Agent will actually send `none` as the env tag. So we could rephrase this to:

```
If this option is left unset or set to an empty value, the environment tag will instead be inherited from the first config option that's actually set to a non-empty value (if any) in this list:
1. top-level `env` option.
2. `env:` tag under the top-level `tags` option.
```

File: pkg/trace/config/apply.go [hunk 10]
```diff
@@ -175,6 +175,18 @@ func (c *AgentConfig) applyDatadogConfig() error {
...
 		c.DefaultEnv = config.Datadog.GetString("apm_config.env")
+		log.Debugf("Setting DefaultEnv to %q (from apm_config.env)", c.DefaultEnv)
+	} else if config.Datadog.IsSet("env") {
+		c.DefaultEnv = config.Datadog.GetString("env")
+		log.Debugf("Setting DefaultEnv to %q (from 'env' config option)", c.DefaultEnv)
+	} else if config.Datadog.IsSet("tags") {
+		for _, tag := range config.Datadog.GetStringSlice("tags") {
+			if strings.HasPrefix(tag, "env:") {
+				c.DefaultEnv = strings.TrimPrefix(tag, "env:")
+				log.Debugf("Setting DefaultEnv to %q (from host tag: %q)", c.DefaultEnv, tag)
```
Comment (olivielpeau) [inline pkg/trace/config/apply.go]: To be more explicit, might make sense to mention the actual config option we pulled this from, i.e.:

```suggestion
				log.Debugf("Setting DefaultEnv to %q (from `env:` entry under the 'tags' config option: %q)", c.DefaultEnv, tag)
```

**Review by olivielpeau — DISMISSED** [review]
LGTM, thanks, only left a few documentation nits, ready to merge for me once my comments are addressed

File: pkg/config/config_template.yaml (hunk 4 above)
Comment (hush-hush) [reply pkg/config/config_template.yaml]: I edited the description

**Review by olivielpeau — APPROVED** [review]
